### PR TITLE
[CSL-2424]  Implement WalletActiveLayer for addresses

### PR DESCRIPTION
**Pull request by ksaric**
https://iohk.myjetbrains.com/youtrack/issue/CSL-2424

We first need to merge https://github.com/input-output-hk/cardano-sl/pull/2758 in order for this to be reviewed and merged.

### CODE REVIEW

**Comment by parsonsmatt** [comment]
I've changed the base to CSL-2423 branch, so that we can more effectively review this now :smile:

File: wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs [hunk 1]
```diff
@@ -14,39 +18,54 @@ module Cardano.Wallet.WalletLayer.Types
...
 ------------------------------------------------------------
 -- Passive wallet layer
 ------------------------------------------------------------
 
+-- The response for the @WalletPassiveLayer@.
+type WalletLayerResponse m a = m (Either WalletLayerError a)
+
+-- A utility function for transforming between representations.
+monadThrowToEither :: forall m a. (MonadCatch m) => m a -> WalletLayerResponse m a
+monadThrowToEither = try
```
Comment (parsonsmatt) [inline wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs]: When we're working with the `PassiveWalletLayer`, do we expect every single function to potentially fail?

When I was implementing the HTTP client, every request is going over the network, which does have an inherent failure possibility. I may be mistaken, but these functions are local, and that suggests to me that they should mostly succeed, except with specific error modes as reflected in the type.

**Comment by SebastienGllmt** [comment]
If possible let's avoid empty names for PRs :)

File: wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs [hunk 2]
```diff
@@ -128,21 +139,21 @@ pwlGetWalletIds = do
     ws          <- askWalletSnapshot
     migrate $ getWalletAddresses ws
 
+
 pwlGetWallet
     :: forall ctx m. (MonadLegacyWallet ctx m)
     => WalletId
-    -> m (Maybe Wallet)
```
[the remark this answers is not in the document]
Comment (ksaric) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: That was my idea as well, but others may not agree - https://github.com/input-output-hk/cardano-sl/pull/2758#pullrequestreview-111661922
In the end, the result will be "something went wrong and you can't get that back" anyway.

File: wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs [hunk 3]
```diff
@@ -113,7 +124,7 @@ pwlCreateWallet NewWallet{..} = do
     wId         <- migrate $ cwId wallet
 
     -- Get wallet or throw if missing.
-    maybeThrow (WalletNotFound wId) =<< pwlGetWallet wId
+    pwlGetWallet wId
```
[the remark this answers is not in the document]
Comment (ksaric) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: It's possible. Note that `MonadLegacyWallet ctx m` contains `MonadThrow m`.

File: wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs [hunk 4]
```diff
@@ -1,6 +1,10 @@
+{-# LANGUAGE StrictData #-}
```
[the remark this answers is not in the document]
Comment (ksaric) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs]: Literally copied from you.

File: wallet/src/Pos/Wallet/Web/State/Storage.hs [hunk 5]
```diff
@@ -389,6 +390,10 @@ getAccountAddrMaps accId = do
   where
     getMap aiLens = fmap (fromMaybe mempty) $ preview $ wsAccountInfos . ix accId . aiLens
 
+-- | A simple way to fetch all the account @CAddress@.
+getAccountAddresses :: WebTypes.AccountId -> Query (CAddresses)
```
[the remark this answers is not in the document]
Comment (ksaric) [reply wallet/src/Pos/Wallet/Web/State/Storage.hs]: This may seem silly, but the reasoning is If we ever change something with the `CAddresses` we don't have to add other parentheses. Which sounds mad when I actually write it down, I can fix this.

File: wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs (hunk 1 above)
Comment (ksaric) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs]: It's reading from a file, so I think we can say it's not really stable and none of them are safe.

**Comment by ksaric** [comment]
@SebastienGllmt Fixed.

File: wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs (hunk 1 above)
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Types.hs]: Sounds good :)

File: wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: `impossible` was a poor word choice :D Is there an error that we can return from this function that is meaningful and potentially expected? eg are there well-typed inputs that predictably cause this function to throw?

File: wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: I'd be happy with `Either WalletNotFound Wallet`, especially if `WalletNotFound` had more information than a single constructor. I don't think the operation should use `MonadThrow` to communicate that.

File: wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ksaric) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: > are there well-typed inputs that predictably cause this function to throw?

You mean something like a wrong password?

[the remark this answers is not in the document]
Comment (parsonsmatt) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: Yeah. `Map.insert` can't fail. But a SQL database insert might fail if a PK collision happens. And in the integration tests, we've already seen a wallet creation fail because the mnemonic was already taken. Returning `Either WalletFail Wallet` would be friendlier as the core part of the API (and a re-throwing variant can be added for convenience)

File: wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs [hunk 6]
```diff
@@ -222,19 +236,73 @@ pwlUpdateAccount wId accIdx accUpdate = do
...
 
-pwlGetAddresses :: WalletId -> m [Address]
-pwlGetAddresses = error "Not implemented!"
+
+pwlCreateAddress
+    :: forall ctx m. (MonadLegacyWallet ctx m)
+    => NewAddress
+    -> m WalletAddress
+pwlCreateAddress NewAddress{..} = do
+    let password = fromMaybe emptyPassphrase $ coerce newaddrSpendingPassword
```
[the remark this answers is not in the document]
Comment (ksaric) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: @KtorZ I don't understand, why would we define a user error here? What error would the user get for what input?

[the remark this answers is not in the document]
Comment (akegalj) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: Ah, I have missread the code. 

Its not clear from description of yt task what `WalletActiveLayer` will be used for? Will `pwlCreateAddress` be called by end user - and `NewWallet` will be passed by him (ie, via http body) ? 

If ^ is the case then I believe address can be created in a password if wallet is not using any password. I am not sure is `newaddrSpendingPassword == Nothing` the same thing as `emptyPassphrase` from `crypto/Pos/Crypto/Signing/Types/Safe.hs`. If it is then the above line might look like:
```haskell
    let password = fromMaybe mempty $ coerce newaddrSpendingPassword
```

In simple terms: "If password is not given from the user, use empty passphrase".

subquestion is, why is `newaddrSpendingPassword` in `NewAddress` modeled like `Maybe SpendingPassword` when `SpendingPassword` already has ability to capture empty passphrase. Maybe I just don't understand distinction between empty passphrase and one not given. Would be happy if anyone can clear this for me.

[the remark this answers is not in the document]
Comment (akegalj) [reply wallet-new/src/Cardano/Wallet/WalletLayer/Legacy.hs]: Ups, I just figured out I just wrote above the same thing you already covered:
```
let password = fromMaybe mempty $ coerce newaddrSpendingPassword
```
mempty == emptyPassphrase - in this case. Sorry for the confusion
